lightningd/watch.c

```diff
@@ -112,6 +112,24 @@ struct txwatch *watch_txid_(const tal_t *ctx,
 	return w;
 }
 
+struct txwatch *find_txwatch(struct chain_topology *topo,
+			     const struct bitcoin_txid *txid,
+			     const struct peer *peer)
+{
+	struct txwatch_hash_iter i;
+	struct txwatch *w;
+
+	/* We could have more than one peer watching same txid, though we
+	 * don't for onchaind. */
+	for (w = txwatch_hash_getfirst(&topo->txwatches, txid, &i);
+	     w;
+	     w = txwatch_hash_getnext(&topo->txwatches, txid, &i)) {
+		if (w->peer == peer)
+			break;
+	}
+	return w;
```

Comment:
This always returns a `struct txwatch` even if it doesn't match the criteria, i.e., if the last one checked doesn't match the `peer` it still gets returned. That is not what [this check](https://github.com/ElementsProject/lightning/pull/833/commits/42073d4874733ea8aee6ebda783d4a0c5d7fe171#diff-b34700033c5cd47ee522323329a7a789R1207) assumes. Please return `NULL` if the watch cannot be found.

Reply:
No, loop terminates on two conditions: w being NULL, or w->peer == peer.  The w != NULL check is buried in the loop conditional...